src/Subscriptions.tsx

```diff
@@ -0,0 +1,134 @@
+import * as React from 'react';
+import * as PropTypes from 'prop-types';
+import ApolloClient, { ApolloError } from 'apollo-client';
+import { DocumentNode } from 'graphql';
+import { ZenObservable } from 'zen-observable-ts';
+import { OperationVariables } from './types';
+
+const shallowEqual = require('fbjs/lib/shallowEqual');
+const invariant = require('invariant');
+
+export interface SubscriptionResult<TData = any> {
+  loading: boolean;
+  data?: TData;
+  error: ApolloError;
+}
+
+export interface SubscriptionProps {
+  query: DocumentNode;
+  variables?: OperationVariables;
+  children: (result: any) => React.ReactNode;
+}
+
+export interface SubscriptionState<TData = any> {
+  loading: boolean;
+  data?: TData;
+  error?: ApolloError;
+}
+
+class Subscription<TData = any> extends React.Component<
+  SubscriptionProps,
+  SubscriptionState<TData>
+> {
+  static contextTypes = {
+    client: PropTypes.object.isRequired,
+  };
+
+  private client: ApolloClient<any>;
+  //TODO: What is the type here?
+  private queryObservable: any;
+  private querySubscription: ZenObservable.Subscription;
+
+  constructor(props: SubscriptionProps, context: any) {
+    super(props, context);
+
+    invariant(
+      !!context.client,
+      `Could not find "client" in the context of Subscription. Wrap the root component in an <ApolloProvider>`,
+    );
+    this.client = context.client;
+    this.initialize(props);
+    this.state = this.getInitialState(props);
+  }
+
+  componentDidMount() {
+    this.startSubscription();
+  }
+
+  componentWillReceiveProps(nextProps, nextContext) {
+    if (
+      shallowEqual(this.props, nextProps) &&
+      this.client === nextContext.client
+    ) {
+      return;
+    }
+
+    if (this.client !== nextContext.client) {
+      this.client = nextContext.client;
+    }
+
+    this.endSubscription();
+    this.initialize(nextProps);
+    this.startSubscription();
+    this.setState(this.getInitialState(nextProps));
+  }
```

Comment:
My preference in this section is that it is unnecessarily broken up.  `initialize` and `getInitialState` are not reused, so i would simply bring their implementation up here.  It increases comprehension without increasing complexity.

I could even say the same about start/end subscription - but I would have to see it before making a judgement.

My go-to in this case would probably be all that code right here.  All init code, not reused, fairly simple, and if needed, a one-line comment to break it up.

Reasoning: when I see code in a method that is one-line I think - _this must be reused, where is it reused and why?_

Reply:
I'm not sure I understand? All of these methods are being reused in the component

Reply:
ah, ok good.  It would have been easier for me to spot in my IDE, but you are right, these are needed.